src/core/tools/ToolFactory.ts

```diff
@@ -0,0 +1,109 @@
+import { BaseTool } from "./BaseTool"
+import { ReadFileTool } from "./implementations/ReadFileTool"
+import { WriteToFileTool } from "./implementations/WriteToFileTool"
+import { ListFilesTool } from "./implementations/ListFilesTool"
+import { SearchFilesTool } from "./implementations/SearchFilesTool"
+import { ExecuteCommandTool } from "./implementations/ExecuteCommandTool"
+import { BrowserActionTool } from "./implementations/BrowserActionTool"
+import { ApplyDiffTool } from "./implementations/ApplyDiffTool"
+import { InsertContentTool } from "./implementations/InsertContentTool"
+import { SearchAndReplaceTool } from "./implementations/SearchAndReplaceTool"
+import { ListCodeDefinitionNamesTool } from "./implementations/ListCodeDefinitionNamesTool"
+import { UseMcpToolTool } from "./implementations/UseMcpToolTool"
+import { AccessMcpResourceTool } from "./implementations/AccessMcpResourceTool"
+import { AskFollowupQuestionTool } from "./implementations/AskFollowupQuestionTool"
+import { SwitchModeTool } from "./implementations/SwitchModeTool"
+import { AttemptCompletionTool } from "./implementations/AttemptCompletionTool"
+import { NewTaskTool } from "./implementations/NewTaskTool"
+import { FetchInstructionsTool } from "./implementations/FetchInstructionsTool"
+import { ToolUseName } from "../assistant-message"
+
+/**
+ * Factory class for creating tool instances
+ */
+export class ToolFactory {
+	private static instance: ToolFactory
+	private toolInstances: Map<string, BaseTool> = new Map()
+
+	/**
+	 * Get the singleton instance of ToolFactory
+	 *
+	 * @returns The ToolFactory instance
+	 */
+	public static getInstance(): ToolFactory {
+		if (!ToolFactory.instance) {
+			ToolFactory.instance = new ToolFactory()
+		}
+		return ToolFactory.instance
+	}
+
+	/**
+	 * Private constructor to enforce singleton pattern
+	 */
+	private constructor() {
+		// Initialize tool instances
+		this.registerTools()
+	}
+
+	/**
+	 * Register all available tools
+	 */
+	private registerTools(): void {
+		// Register all tools
+		this.registerTool(new ReadFileTool())
+		this.registerTool(new WriteToFileTool())
+		this.registerTool(new ListFilesTool())
+		this.registerTool(new SearchFilesTool())
+		this.registerTool(new ExecuteCommandTool())
+		this.registerTool(new BrowserActionTool())
+		this.registerTool(new ApplyDiffTool())
+		this.registerTool(new InsertContentTool())
+		this.registerTool(new SearchAndReplaceTool())
+		this.registerTool(new ListCodeDefinitionNamesTool())
+		this.registerTool(new UseMcpToolTool())
+		this.registerTool(new AccessMcpResourceTool())
+		this.registerTool(new AskFollowupQuestionTool())
+		this.registerTool(new SwitchModeTool())
+		this.registerTool(new AttemptCompletionTool())
+		this.registerTool(new NewTaskTool())
+		this.registerTool(new FetchInstructionsTool())
+	}
```

Comment:
Can we make tool register to this factory by itself (we can do it in base class), and this class look like a registry than a factory